src/type-checker.c

```diff
@@ -0,0 +1,523 @@
+/*
+ * Copyright 2017 WebAssembly Community Group participants
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "type-checker.h"
+
+#define CHECK_RESULT(expr) \
+  do {                     \
+    if (WABT_FAILED(expr)) \
+      return WABT_ERROR;   \
+  } while (0)
+
+#define COMBINE_RESULT(result_var, result) \
+  do {                                     \
+    result_var |= (result);                \
+  } while (0)
+
+static void WABT_PRINTF_FORMAT(2, 3)
+    print_error(WabtTypeChecker* tc, const char* fmt, ...) {
+  if (tc->error_handler->on_error) {
+    WABT_SNPRINTF_ALLOCA(buffer, length, fmt);
+    tc->error_handler->on_error(buffer, tc->error_handler->user_data);
+  }
+}
+
+WabtResult wabt_typechecker_get_label(WabtTypeChecker* tc,
+                                      size_t depth,
+                                      WabtTypeCheckerLabel** out_label) {
+  if (depth >= tc->label_stack.size) {
+    assert(tc->label_stack.size > 0);
+    print_error(tc, "invalid depth: %" PRIzd " (max %" PRIzd ")",
+                depth, tc->label_stack.size - 1);
```

Comment:
Would it make sense to set out_label to NULL here? That would guarantee that if the caller inadvertently used the pointer, it would fail (rather than potentially using some out of date value).